File: src/main/java/com/bazbatlabs/bricksmash/models/Rect.java
package com.bazbatlabs.bricksmash.models;

public final class Rect {

    public final Vec2 origin;
    public final Vec2 size;

    public Rect(Vec2 origin, Vec2 size) {
        this.origin = origin;
        this.size = size;
    }

    public float left() { return origin.x; }
    public float right() { return origin.x + size.x; }
    public float bottom() { return origin.y; }
    public float top() { return origin.y + size.y; }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Rect)) {
            return false;
        }
        Rect that = (Rect)o;

        return this.origin == that.origin && this.size == that.size;
    }

    @Override
    public int hashCode() {
        int result = 17;
        result = 31 * result + origin.hashCode();
        result = 31 * result + size.hashCode();
        return result;
    }

    @Override
    public String toString() {
        return String.format("(x: %.4f, y: %.4f, w: %.4f, h: %.4f)",
                             origin.x, origin.y, size.x, size.y);
    }
}

File: src/main/scala/lib/Rect.scala
package com.bazbatlabs.bricksmash.lib

class Rect(val origin: Vec2, val size: Vec2) {

  def left = origin.x
  def right = origin.x + size.x
  def bottom = origin.y
  def top = origin.y + size.y

  override def equals(other: Any): Boolean =
    other match {
      case that: Rect =>
        (that canEqual this) && origin == that.origin && size == that.size

      case _ => false
    }

  def canEqual(other: Any): Boolean = other.isInstanceOf[Rect]

  override def hashCode: Int = (31 * (31 + origin.hashCode)) + size.hashCode

  override def toString = f"x: ${origin.x}%.4f, y: ${origin.y}%.4f, " +
    f"w: ${size.x}%.4f, h: ${size.y}%.4f"
}

object Rect {
  def apply(origin: Vec2, size: Vec2) = new Rect(origin, size)
}
